MDANSE/Src/MDANSE/MolecularDynamics/Trajectory.py

```diff
@@ -66,88 +67,91 @@ def __init__(self, filename, trajectory_format=None):
-                self._atom_names = copy.deepcopy(self.atom_types)
-                for k, v in temp_names.items():
-                    self._atom_names[k] = v
-        return self._atom_names
+        return temp
 
-    @property
+    @cached_property
     def element_from_label(self) -> dict[str, str]:
-        """Mapping from atom label to its chemical element name."""
-        if not self._element_from_label:
-            self.atom_names
-        return self._element_from_label
-
-    @property
+        """Maps the full atom labels to the chemical elements.
+        
+        If grouping is used, atom labels may contain the molecule name
+        as well as the chemical element of the atom, and will not match
+        the entries in the atom database. This dictionary allows to get
+        a valid atom database key for each atom in the system.
+        """
+        mapping = {element : element for element in self.unique_elements}
+        if self._grouping_level == "molecule":
+            temp_names = {}
+            for mol_name in self.chemical_system._clusters:
+                for cluster in self.chemical_system._clusters[mol_name]:
+                    overlap = set(cluster).intersection(self.atom_indices)
+                    for x in overlap:
+                        temp_names[f"<{mol_name}>/{self.atom_types[x]}"] = self.atom_types[x]
+            mapping.update(temp_names)
+        return mapping
+    
+    @cached_property
     def group_lookup(self) -> dict[str, int] | dict[str, list[int]]:
         """Dictionary of currently existing groups.
         
         The keys are names of the group. The values can be the count
         of all atoms belonging to the group, or (for 'each molecule' only)
         a list of atom indices belonging to the individual molecule.
         """
-        if not self._group_lookup:
-            self.atom_names
+        temp_dict = {}
         if self._grouping_level == "each molecule":
-            return {k: v for k, v in self._group_lookup.items() if len(v) > 0}
-        return {k: v for k, v in self._group_lookup.items() if v > 0}
+            for mol_name, clusters in self.chemical_system._clusters.items():
+                for mol_number, cluster in enumerate(clusters):
+                    if set(cluster).issubset(self.atom_indices):
+                        temp_dict[f"{mol_name}_mol{mol_number + 1}"] = cluster
+        elif self._grouping_level == "molecule":
+            for mol_name in self.chemical_system._clusters:
+                temp_dict.setdefault(mol_name, 0)
+                for mol_number, cluster in enumerate(
+                    self.chemical_system._clusters[mol_name],
+                ):
+                    overlap = set(cluster).intersection(self.atom_indices)
+                    if not overlap:
+                        continue
+                    temp_dict[mol_name] += len(overlap)
```

Comment:
`if not overlap` the `len` will be `0`, so may just want to remove the `if` for simpler code?